firmware-support/bittide-hal-c/build.rs

```diff
@@ -3,60 +3,303 @@
+                &mut pass,
+                std::iter::once(reg.type_ref),
+            );
+        }
+    }
+
+    for hal in &deduped_hals {
+        for dev in &hal.devices {
+            let desc = &ctx.device_descs[*dev];
+            for reg in &ctx.registers[desc.registers] {
+                monomorpher.monomorph_toplevel_type_refs(
+                    &mut varis,
+                    &mut pass,
+                    std::iter::once(reg.type_ref),
+                );
+            }
         }
+    }
+
+    monomorpher.monomorph_type_descs(&mut varis, &mut pass);
+
+    // we're going for a folder structure like this
+    //
+    // generated/
+    //   types/
+    //   shared_devices/
+    //   hals/
+    //     <hal-name>/
+    //       devices/
+    //     <hal-name>.h  <- contains the device instances
 
-        let stem = path
-            .file_stem()
-            .and_then(|s| s.to_str())
-            .expect("Failed to get file stem");
+    let src_path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("generated");
+    let types_path = src_path.join("types");
+    let shared_devices_path = src_path.join("shared_devices");
 
-        println!("cargo:rerun-if-changed={}", path.display());
+    // clear out generated shared code first
+    _ = std::fs::remove_dir_all(&shared_devices_path);
+    _ = std::fs::remove_dir_all(&types_path);
 
-        let src = fs::read_to_string(&path)
-            .unwrap_or_else(|e| panic!("Failed to read {}: {e}", path.display()));
+    {
+        // types
 
-        let desc = parse(&src).unwrap_or_else(|e| panic!("Failed to parse memory map: {e}"));
+        std::fs::create_dir_all(&types_path)
+            .expect("Create (gitignored) `generated/types` directory");
 
-        memory_maps.insert(stem.to_string(), desc);
+        for handle in &shared.deduped_types {
+            if ctx.type_primitives.contains(handle) {
+                continue;
+            }
+            if ctx.type_tuples.contains(handle) {
+                continue;
+            }
+            let (name, code, refs) =
+                backend_c::type_desc::generate_type_desc(&ctx, &varis, *handle);
+            let mod_name = ident(IdentType::Module, type_name(name));
+
+            let file_path = types_path.join(format!("{mod_name}.h"));
+            let mut file = File::create(&file_path).unwrap();
+            with_guard(
+                &mut file,
+                &ident(IdentType::Define, format!("types_{mod_name}_H")),
+                |file| {
+                    writeln!(file, "{}", gen_imports()).unwrap();
+
+                    for ty in &refs.references {
+                        let name = &ctx.type_names[*ty];
+                        let mod_name = ident(IdentType::Module, &name.base);
+                        writeln!(file, "#include \"types/{}.h\"", mod_name).unwrap();
+                    }
+
+                    // writeln!(file, "{}", generate_type_ref_imports(&ctx, &refs)).unwrap();
```

Comment:
Why is that `generate_type_ref_import` commented out, or rather, why is it done inline?